feat(ma): adding tiktok ads source integration

jabahamondes opened the pull request:
## Problem

Adding tiktok integration

## Changes

- adding OAuth integration, tik tok use specific logic for almost everything: getting the access token, refreshing,
- defining the schema in settings.py
- creating typical source file
- creating a specific tiktok paginator
- adding exponential backoff when doing the api call (when getting the resource), this decorator has specific logic for tiktok codes
- creating some utility functions

## How did you test this code?
Added unit tests and manually:
First sync:
<img width="1297" height="408" alt="image" src="https://github.com/user-attachments/assets/22f1a8cd-7561-4ab6-bbf5-91d3ce2155f8" />

<Added a new campaign>
Incremental sync: (paid for a new campaign and wait it for next day to be synced)
<img width="1259" height="444" alt="image" src="https://github.com/user-attachments/assets/368e00fa-d37e-4300-ab26-45c0b859447f" />

Commit:
esqueleton tik tok ads

Commit:
fix: correct TikTok ad naming in schema and types (#38610)

Commit:
Merge conflicts

github-actions[bot] committed:
Update query snapshots

Commit:
Merge conflicts

Commit:
Merge branch 'feat/adding-tiktok-ads-integration-squeleton' of github.com:PostHog/posthog into feat/adding-tiktok-ads-integration-squeleton

Commit:
feat(ma): Implement TikTok Ads integration with OAuth support and data fetching

- Added TikTok Ads as a new integration in the system.
- Implemented OAuth flow for TikTok Ads, including credential validation and token management.
- Created settings and configuration for TikTok Ads endpoints.
- Developed utility functions for flattening TikTok report data and handling pagination.
- Added tests for the TikTok Ads source and utility functions to ensure functionality and correctness.
- Included date chunking logic to comply with TikTok's API limits.

This integration allows users to collect campaign data and advertising metrics from TikTok Ads, enhancing our data import capabilities.

Commit:
fixing mypy baseline errors

Commit:
Merge conflicts

Commit:
Removing repeated file

Commit:
Adding exponencial backoff and fixing tests

Commit:
Refactor TikTok resource handling to use template placeholders for date parameters

- Removed start_date and end_date parameters from get_tiktok_resource function.
- Updated tests to assert that date parameters are now represented as template placeholders.
- Adjusted resource fetching logic to align with the new parameter structure.

Commit:
renaming file

lricoy committed:
Merge branch 'master' into feat/adding-tiktok-ads-integration

lricoy reviewed (approved):
lgtm, just some comments!

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-tiktok-ads-integration

Commit:
feat: implement async retry logic with exponential backoff

- Added  utility for async functions in .
- Updated TikTok Ads integration to use the new retry logic in .
- Removed deprecated  function and related tests.
- Cleaned up unnecessary comments and improved exception handling in TikTok Ads utilities.

Commit:
refactor: update TikTok Ads utilities for improved structure and functionality

- Replaced deprecated flattening functions with methods in TikTokReportResource.
- Introduced TikTokDateRangeManager for date range calculations and chunking.
- Updated tests to reflect new utility methods and ensure proper functionality.
- Cleaned up unused imports and improved error handling in TikTok Ads integration.

Commit:
Moving test files inside test folder

Commit:
Reverting moving utility function for retries

Commit:
updating retryable codes

Commit:
Removing is report fn and put that logic in the endpoint config

Commit:
Supporting append

Commit:
Removing async retry and creating a new sync one

Commit:
Refactor TikTok Ads resource retrieval to use inline lambda for retry logic

- Removed the separate  function and integrated its logic directly into the retryable call.
- Simplified the retry mechanism for resource creation with a lambda function, maintaining TikTok-specific retry settings.

Commit:
refactor: improve record flattening logic in TikTokReportResource

- Simplified the flattening process by checking for metrics and dimensions separately.
- Removed unnecessary else clause for cleaner code structure.
- Ensured that the function returns the original record if no flattening occurs.

Commit:
update with more error codes

Commit:
refactor: update TikTokAdsSource and TikTokReportResource for clarity and functionality

- Set supports_append to False in TikTokAdsSource to reflect current capabilities.
- Simplified process_resources method in TikTokReportResource to improve readability and efficiency.
- Removed unnecessary methods for resource flattening, consolidating logic into a single method for better maintainability.

Commit:
refactor: enhance TikTokAds settings and report resource transformations

- Updated MAX_TIKTOK_DAYS_TO_QUERY to 29 to comply with TikTok's API limits.
- Introduced comprehensive auction fields for TikTok reports, improving data granularity.
- Replaced flattening logic with dedicated transformation methods for insights, management, and advertisers records in TikTokReportResource.
- Added pre_transform method to streamline record processing based on endpoint type.
- Updated tests to reflect changes in transformation logic, ensuring accuracy in data handling.

Commit:
Fixed python type checking

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-tiktok-ads-integration

Commit:
feat(migrations): add integration kind field with updated choices

- Created migration 0880 to alter the 'kind' field in the 'integration' model, expanding the choices to include new integrations like TikTok, Databricks, and more.
- Updated max_migration.txt to reflect the new migration version.

Commit:
feat(config): add advertiser_id and tiktok_integration_id to TikTokAdsSourceConfig

- Introduced  as a required string field and  as an integer field with a conversion from string to int for TikTokAds source configuration.

github-actions[bot] committed:
Update query snapshots

Gilbert09 reviewed (commented):
Just sort out the licensed code and this is good to go

Commit:
refactor(tiktok_ads): enhance settings and transformation logic

- Introduced EndpointType enum to categorize endpoint types (ACCOUNT, ENTITY, REPORT) for better clarity and maintainability.
- Replaced is_report_endpoint boolean with endpoint_type in TIKTOK_ADS_CONFIG for improved configurability.
- Updated transformation methods in TikTokReportResource to utilize endpoint_type, ensuring correct processing based on endpoint category.
- Enhanced tests to validate new transformation logic and ensure accurate handling of TikTok reports.

Commit:
Fixing test

Commit:
Merge branch 'feat/adding-tiktok-ads-integration' of github.com:PostHog/posthog into feat/adding-tiktok-ads-integration

Commit:
Fixing test

Commit:
Merge conflicts

jabahamondes commented on posthog/temporal/data_imports/sources/tiktok_ads/test/test_tiktok_ads_utils.py:
```diff
@@ -0,0 +1,745 @@
+
+import pytest
+from unittest.mock import Mock
+
```
```suggestion
from enum import Enum
```

Commit:
Fixing tests

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-tiktok-ads-integration

Commit:
Fixing tests
